src/sfctl/custom_app.py

```diff
@@ -179,12 +183,238 @@ def upload_to_native_imagestore(sesh, endpoint, abspath, basename, #pylint: disa
+class IgnoreCopy():  # pylint: disable=too-few-public-methods
+    """
+    A class which contains information and methods for the shutil.copytree method's
+    callback parameter.
+    """
+
+    # Directories which will be ignored as part of the ignore_copy function
+    # If used for compressing application package:
+    # A list of strings representing the abs path of the dirs in an application package
+    # which need to be compressed
+    # Paths in dirs_to_ignore should normalized using the _normalize_path function before setting
+    dirs_to_ignore = []
+
+    def ignore_copy(self, directory_being_visited, list_of_dirs):
+        """
+        The ignore function for shutil.copytree()
+
+        :param directory_being_visited: str
+        :param list_of_dirs: Example: ['t.txt']
+
+        :return: a subset of the items in its second argument (must be relative path).
+                 these names will then be ignored in the copy process
+        """
+
+        to_ignore = []
+
+        for directory in list_of_dirs:
+
+            full_path = os.path.join(directory_being_visited, directory)
+            full_path = _normalize_path(full_path)
+            if full_path in self.dirs_to_ignore:
+                to_ignore.append(directory)
+
+        return to_ignore
+
+def _normalize_path(path):
+    """
+    Standardize a path to a file/folder location so that they can be compared
+
+    :param path: (str) represents a path on a machine
+    :return: (str) the standardized path
+    """
+
+    path = os.path.realpath(path)  # This removes slashes at the end of the path
+    path = os.path.normpath(path)
+    path = os.path.normcase(path)
+    path = os.path.abspath(path)
+    return path
+
+def compress_package(app_dir, output_dir):
+    """
+    Compress to the location passed in (output_dir). Note that it is not the entire package
+    which is compressed, but rather, only some inside parts of the app package folder.
+
+    Check if the folder has the correct structure for a service fabric application. If
+    not, raise an exception alerting user of bad folder structure.
+
+    ZIP64 functionality is present for Python
+    versions 3.4 and above.
+
+    For example, if app_dir = C:/SomeFolder/WordCountApp
+    and if output_dir = C:/SomeLocation,
+    then the following will be created: C:/SomeLocation/WordCountApp
+
+    :param app_dir: (str) An absolute path to an application package to be compressed
+
+    :param output_dir: (str) An absolute path to the location to output the zipped dir
+
+    :return: Nothing, or a CLIError exception
+    """
+
+    # Check if we're dealing with a dir in _check_folder_structure_and_get_dirs instead of
+    # in this function
+
+    # Normalize slashes, etc, in app_dir and output_dir
+    app_dir = _normalize_path(app_dir)
+    output_dir = _normalize_path(output_dir)
+
+    compress_copy = IgnoreCopy()
```

Comment:
If the input pkg is compressed, then we shouldn't throw error. Just copy to the output location.